packages/components/src/truncate/truncate-utils.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * External dependencies
+ */
+import { isNil } from 'lodash';
+
+export const TRUNCATE_ELLIPSIS = '…';
+export const TRUNCATE_TYPE = {
+	auto: 'auto',
+	head: 'head',
+	middle: 'middle',
+	tail: 'tail',
+	none: 'none',
+};
+
+export const TRUNCATE_DEFAULT_PROPS = {
+	ellipsis: TRUNCATE_ELLIPSIS,
+	ellipsizeMode: TRUNCATE_TYPE.auto,
+	limit: 0,
+	numberOfLines: 0,
+};
+
+// Source
+// https://github.com/kahwee/truncate-middle
+/**
+ * @param {string} word
+ * @param {number} headLength
+ * @param {number} tailLength
+ * @param {string} ellipsis
+ */
+export function truncateMiddle( word, headLength, tailLength, ellipsis ) {
+	if ( typeof word !== 'string' ) {
+		return '';
+	}
+	const wordLength = word.length;
+	// Setting default values
+	// eslint-disable-next-line no-bitwise
+	const frontLength = ~~headLength; // will cast to integer
+	// eslint-disable-next-line no-bitwise
+	const backLength = ~~tailLength;
+	/* istanbul ignore next */
+	const truncateStr = ! isNil( ellipsis ) ? ellipsis : TRUNCATE_ELLIPSIS;
+
+	if (
+		( frontLength === 0 && backLength === 0 ) ||
+		frontLength >= wordLength ||
+		backLength >= wordLength ||
+		frontLength + backLength >= wordLength
+	) {
+		return word;
+	} else if ( backLength === 0 ) {
+		return word.slice( 0, frontLength ) + truncateStr;
+	}
+	return (
+		word.slice( 0, frontLength ) +
+		truncateStr +
+		word.slice( wordLength - backLength )
```

Comment:
I'm pretty sure  this kind of truncating using substrings is language specific. How does this component behave in other languages? cc @swissspidy

Reply:
Some tests with strings in other languages would be good.

So if I understand this right, `truncateMiddle( 'Hello', 1, 1, '...')` will return `H...o`.

Sounds OK when using Latin alphabet, but with strings like `こんにちは` ("Kon'nichiwa", "Hello" in Japanese) this will result in `こ...は`, which probably is less understandable.

Not sure how it would work with Arabic, but you'd know that best :-)

Worst case we'd need some words/characters logic like the word count package has

Reply:
Interesting 🤔 ! if there's a library folks are aware of that could elegantly handle this, I think it would be a worth while addition :D.

---

For context, So far, we've only used it for `Truncating...` handling.

I based this component on something I've built in the past. Also heavily drawing inspiration from how `Text` and truncation is handled in React Native:

https://reactnative.dev/docs/text#ellipsizemode

Reply:
Adding more thoughts!

Taking a step back, for the use-cases of user interfaces, realistically the only situations were we'd need to `truncateMiddle` would be for long file names.

For example:

`thequickbrownfoxjumpsoverthelazydog.jpg`

maybe becomes

`theq...ydog.jpg`

This is common for attachment related UIs or slug/tags.

In that case, it doesn't necessarily matter if the word/letters is linguistically understandable.
As long as there's enough to identify either the extension (e.g. `.jpg`) or the file / keyword / slug :)

Reply:
Excuse my idiot questions: 

 - if middle is not the important use-case for this component, should we include it?
 - It seems this code has some logic to split words... I believe we may already have some similar logic in `@wordpress/wordcount` that is very similar and that works across languages, If we keep this behavior, do you  think there's a way to kind of consolidate that logic somehow?

Reply:
I think we should include `middle` as long as the use case that @ItsJonQ brought up is one that we'll need available at any time (or if block developers would find it useful for file names).

We don't do any word counting in this component, just line counting. Was there a specific place @youknowriad that you would see us using `@wordpress/wordcount` for this component?

Reply:
I guess I misunderstood the code but I thought there was some smart behavior to try to split between "words"  or something like that in which case, it  would be a similar algorithm to word count to figure out  where are the words.

Reply:
There’s no such logic, as far as I can see. @ItsJonQ maybe you can chime in here to verify.

Reply:
I don't believe there's logic to split words - just splitting characters.